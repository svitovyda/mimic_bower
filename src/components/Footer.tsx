import * as React from "react";

export const Footer: React.FC = () => {
  return <div></div>;
};
