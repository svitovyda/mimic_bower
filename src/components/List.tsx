import type { Package } from '../models/Package';
import { ModuleItem } from './ModuleItem';
import * as React from 'react';

export interface ListProps {
  data: Package[];
}

export const List: React.FC<ListProps> = ({ data }: ListProps) => {
  return (
    <div>
      {data.map((p, i) => (
        <ModuleItem key={`${i}-${p.name}-${p.platform}`} packageElement={p} />
      ))}
    </div>
  );
};
