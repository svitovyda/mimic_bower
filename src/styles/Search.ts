import styled from 'styled-components';

export const InputContainer = styled.div`
  display: block;
`;

export const Input = styled.input`
  display: inline-block;
  padding-block: 1px;
  padding-inline: 2px;
  padding: 12px 10px;
  cursor: text;
  height: 45px;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #ccc;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin-bottom: 10px;
  outline-width: 0;
  color: inherit;
`;
