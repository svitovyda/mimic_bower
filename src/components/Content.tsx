import { useContentFetcher } from "../hooks/contentFetcher";
import { usePagination } from "../hooks/usePagination";
import type { Package } from "../models/Package";
import { List } from "./List";
import { Pagination } from "./Pagination";
import { LoaderAnimation } from "./ui/LoaderAnimation";
import configJson from "config";
import * as React from "react";

const ItemsPerPage = configJson.searchItemsPageMaxSize;

export interface ContentProps {
  query: string;
}

export const Content: React.FC<ContentProps> = ({ query }: ContentProps) => {
  const [fetchState, setQuery] = useContentFetcher("");
  const [paginationState, navigateToPage] = usePagination(fetchState.data);
  const [pageData, setPageData] = React.useState<Package[]>([]);

  React.useEffect(() => {
    setQuery(query);
    setPageData([]);
  }, [query, setQuery]);

  React.useEffect(() => {
    if (fetchState.data.length === 0) setPageData([]);
    else {
      const pageStart = (paginationState.currentPage - 1) * ItemsPerPage;

      setPageData(fetchState.data.slice(pageStart, pageStart + ItemsPerPage));
    }
  }, [fetchState.data, paginationState]);

  if (fetchState.isLoading) return <LoaderAnimation />;
  if (fetchState.isError) return <div>Sorry, there was an error while fetching repositories</div>;
  return (
    <div>
      <List data={pageData} />
      <Pagination
        currentPage={paginationState.currentPage}
        totalPages={paginationState.totalPages}
        navigateToPage={navigateToPage}
      />
    </div>
  );
};
