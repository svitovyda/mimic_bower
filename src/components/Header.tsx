import { HeaderContainer, HeaderLogo, HeaderSubtitle, HeaderTitle } from '../styles/Header';
import * as React from 'react';

export const Header: React.FC = () => {
  return (
    <header>
      <HeaderContainer>
        <HeaderLogo />
        <HeaderTitle>Bower Search</HeaderTitle>
        <HeaderSubtitle>
          Powered by{' '}
          <a href="https://libraries.io/" target="_blank" rel="noreferrer">
            libraries.io
          </a>
        </HeaderSubtitle>
      </HeaderContainer>
    </header>
  );
};
