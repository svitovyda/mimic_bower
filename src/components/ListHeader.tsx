/* eslint-disable jsx-a11y/anchor-is-valid */
import type { SortType } from '../services/ApiServise';
import { ListHeaderContainer, NameLink, OwnerLink, StarsLink } from '../styles/ListHeader';
import * as React from 'react';

export interface ListHeaderProps {
  onSort: (sortBy?: SortType) => void;
}

export const ListHeader: React.FC<ListHeaderProps> = React.memo(({ onSort }: ListHeaderProps) => {
  return (
    <ListHeaderContainer>
      <NameLink onClick={() => onSort()}>Name</NameLink>
      <OwnerLink>Owner</OwnerLink>
      <StarsLink onClick={() => onSort('stars')}>Stars</StarsLink>
    </ListHeaderContainer>
  );
});

ListHeader.displayName = 'ListHeader';
