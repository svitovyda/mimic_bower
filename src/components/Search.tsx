import { Input, InputContainer } from '../styles/Search';
import configJson from 'config';
import debounce from 'debounce';
import * as React from 'react';

export interface SearchProps {
  onQueryChanged: (newQuery: string) => void;
}

const MinAcceptable = configJson.minimumSearchQueryLength || 3;
const DebounseInterval = configJson.searchInputDebounse || 1000;

export const Search: React.FC<SearchProps> = React.memo(({ onQueryChanged }: SearchProps) => {
  const [query, setQuery] = React.useState<string>('');

  const updateSearchQuery = debounce((value) => {
    if (value.length > MinAcceptable) {
      onQueryChanged(value);
    }
  }, DebounseInterval);

  const onQueryChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;
    setQuery(newValue);
    updateSearchQuery(newValue);
  };

  return (
    <InputContainer>
      <Input
        value={query}
        maxLength={100}
        onChange={onQueryChange}
        autoComplete="off"
        type="search"
        placeholder="Search..."
      />
    </InputContainer>
  );
});

Search.displayName = 'Search';
