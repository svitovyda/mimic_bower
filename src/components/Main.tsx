import { MainContainer, RightContainer } from "../styles/Main";
import { Content } from "./Content";
import { Footer } from "./Footer";
import { Search } from "./Search";
import { Sidebar } from "./Sidebar";
import * as React from "react";

export const Main: React.FC = () => {
  const [query, setQuery] = React.useState<string>("");

  const onNewQuery = React.useCallback(
    (newQuery: string) => {
      setQuery(newQuery);
    },
    [setQuery]
  );

  return (
    <MainContainer>
      <Sidebar />
      <RightContainer>
        <Search onQueryChanged={onNewQuery} />
        <Content query={query} />
        <Footer />
      </RightContainer>
    </MainContainer>
  );
};
