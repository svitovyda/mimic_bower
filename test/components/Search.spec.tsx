import { Search } from "../../src/components/Search";
import { render } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";

describe("Search", () => {
  it("renders without error", () => {
    const component = render(<Search onQueryChanged={jest.fn()} />);

    expect(component).toMatchSnapshot();
  });
});
