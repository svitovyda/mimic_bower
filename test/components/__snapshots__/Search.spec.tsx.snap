// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c0 {
  display: block;
}

.c1 {
  display: inline-block;
  padding-block: 1px;
  padding-inline: 2px;
  padding: 12px 10px;
  cursor: text;
  height: 45px;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #ccc;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin-bottom: 10px;
  outline-width: 0;
  color: inherit;
}

<body>
    <div>
      <div
        class="c0"
      >
        <input
          autocomplete="off"
          class="c1"
          maxlength="100"
          placeholder="Search..."
          type="search"
          value=""
        />
      </div>
    </div>
  </body>,
  "container": .c0 {
  display: block;
}

.c1 {
  display: inline-block;
  padding-block: 1px;
  padding-inline: 2px;
  padding: 12px 10px;
  cursor: text;
  height: 45px;
  border: 0;
  border-radius: 1px;
  box-shadow: 0 0 1px #ccc;
  font-size: 18px;
  font-weight: 400;
  width: 100%;
  margin-bottom: 10px;
  outline-width: 0;
  color: inherit;
}

<div>
    <div
      class="c0"
    >
      <input
        autocomplete="off"
        class="c1"
        maxlength="100"
        placeholder="Search..."
        type="search"
        value=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
